src/idom/utils.py

```diff
@@ -63,80 +67,141 @@ def html_to_vdom(source: str, *transforms: _ModelTransform) -> Dict[str, Any]:
-                        child = t(child)
-                if child is not None:
-                    transformed.append(child)
-                    to_visit.append(child)
-            node["children"] = transformed
-            if "attributes" in node and not node["attributes"]:
-                del node["attributes"]
-            if "children" in node and not node["children"]:
-                del node["children"]
-    return root
-
-
-class HtmlParser(_HTMLParser):
-    """HTML to VDOM parser
-
-    Example:
-
-        .. code-block::
-
-            parser = HtmlParser()
-
-            parser.feed(an_html_string)
-            parser.feed(another_html_string)
-            ...
-
-            vdom = parser.model()
+    if not isinstance(html, str):
+        raise TypeError(f"Encountered unsupported type {type(html)} from {html}")
+
+    # If the user provided a string, convert it to a list of lxml.etree nodes
+    parser = etree.HTMLParser(
+        remove_comments=True,
+        remove_pis=True,
+        remove_blank_text=True,
+        recover=False,
+    )
+    nodes: List = fragments_fromstring(html, no_leading_text=True, parser=parser)
+    has_root_node = len(nodes) == 1
+
+    # Find or create a root node
+    if has_root_node:
+        root_node = nodes[0]
+    else:
+        root_node = etree.Element("div", None, None)
+        for child in nodes:
+            root_node.append(child)
+
+    # Convert the lxml node to a VDOM dict
+    vdom = _etree_to_vdom(root_node, transforms)
+
+    # Change the artificially created root node to a React Fragment, instead of a div
+    if not has_root_node:
+        vdom["tagName"] = ""
+
+    return vdom
+
+
+def _etree_to_vdom(node: etree._Element, transforms: Iterable[_ModelTransform]) -> Dict:
+    """Recusively transform an lxml etree node into a DOM model
+
+    Parameters:
+        source:
+            The ``lxml.etree._Element`` node
+        transforms:
+            Functions of the form ``transform(old) -> new`` where ``old`` is a VDOM
+            dictionary which will be replaced by ``new``. For example, you could use a
+            transform function to add highlighting to a ``<code/>`` block.
     """
+    if not isinstance(node, etree._Element):
+        raise TypeError(f"Encountered unsupported type {type(node)} from {node}")
+
+    # This will recursively call _etree_to_vdom() on all children
+    children = _generate_vdom_children(node, transforms)
+
+    # Convert the lxml node to a VDOM dict
+    attributes = dict(node.items())
+    key = attributes.pop("key", None)
+    vdom = (
+        # If available, use a constructor from idom.html to create the VDOM dict
+        getattr(idom.html, node.tag)(attributes, *children, key=key)
+        if hasattr(idom.html, node.tag)
+        # Fall back to using a generic VDOM dict
+        else {
+            "tagName": node.tag,
+            "children": children,
+            "attributes": attributes,
+            "key": key,
+        }
+    )
```

Comment:
The built-in constructors already prune empty fields so, if you only assign fields whose values are not empty, you don't need `_prune_vdom_fields` anymore.

```suggestion
    attributes = dict(node.items())
    key = attributes.pop("key", None)

    if hasattr(idom.html, node.tag):
        vdom = getattr(idom.html, node.tag)(attributes, *children, key=key)
    else:
        vdom = {"tagName": node.tag}
        if children:
            vdom["children"] = children
        if attributes:
            vdom["attributes"] = attributes
        if key is not None:
            vdom["key"] = key
```

Reply:
This proposed solution isn't using list comprehension. So it will be slower at scale.

Reply:
I don't see a list comprehension in the original implementation, just an inline if statement. Commented in wrong location?

Reply:
Yep